File: 0143_reorder_list/reorder_list.java
public class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

// Time: O(n)
// Space: O(1)
class Solution {
    public void reorderList(ListNode head) {
        if (head == null) {
            return;
        }
        ListNode mid = findMiddleNode(head);
        ListNode midHead = reverseList(mid);
        mergeLists(head, midHead);
    }

    private ListNode findMiddleNode(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    private ListNode reverseList(ListNode head) {
        ListNode prev = null, curr = head, tmp;
        while (curr != null) {
            tmp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = tmp;
        }
        return prev;
    }

    private void mergeLists(ListNode head1, ListNode head2) {
        ListNode first = head1, second = head2, tmp;
        while (second.next != null) {
            tmp = first.next;
            first.next = second;
            first = tmp;
            tmp = second.next;
            second.next = first;
            second = tmp;
        }
    }
}

File: 0143_reorder_list/reorder_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reorder_list(head: ListNode) -> None:
    """
    Do not return anything, modify head in-place instead.
    Time: O(n)
    Space: O(1)
    """
    if not head:
        return
    # find the middle of linked list
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    # reverse the second part of the list
    prev, curr = None, slow
    while curr:
        curr.next, prev, curr = prev, curr, curr.next

    # merge two sorted linked lists
    first, second = head, prev
    while second.next:
        first.next, first = second, first.next
        second.next, second = first, second.next
